icaruscode/Filters/FilterCRTPMTMatching_module.cc

```diff
@@ -0,0 +1,618 @@
+
+// LArSoft includes
+#include "larcore/CoreUtils/ServiceUtil.h"
+#include "lardata/DetectorInfoServices/DetectorClocksService.h"
+#include "larcorealg/Geometry/GeometryCore.h"
+#include "larcore/Geometry/Geometry.h"
+
+// Data product includes
+#include "lardataobj/RecoBase/OpHit.h"
+#include "lardataobj/RecoBase/OpFlash.h"
+#include "sbnobj/Common/CRT/CRTHit.hh"
+#include "icaruscode/CRT/CRTUtils/CRTCommonUtils.h"
+#include "sbnobj/Common/Trigger/ExtraTriggerInfo.h"
+
+// C++ includes
+#include <vector>
+#include <map>
+#include <numeric>
+
+// ROOT includes
+#include "TTree.h"
+#include "TVector3.h"
+
+namespace icarus
+{
+    namespace crt
+    {
+        class FilterCRTPMTMatching;
+        struct CRTPMT
+        {
+            double tof;
+            double distance;
+            art::Ptr<sbn::crt::CRTHit> CRTHit;
+        };
+        struct MatchedCRT
+        {
+            // vector of pairs where first is the matched Time of Flight and second is the matched CRTHit
+            std::vector<CRTPMT> entering;
+            std::vector<CRTPMT> exiting;
+            int matchType;
+        };
+        struct TrajPoint
+        {
+            double X;
+            double Y;
+            double Z;
+            double T;
+        };
+    }
+}
+
+using namespace icarus::crt;
+
+bool flashInTime(double const &flashTime, int gateType, double gateDiff, double flashTimeCut)
+{
+    std::map<unsigned int, double> gateLength{
+        {1, 2.2},  // BNB
+        {2, 10.1}, // NuMI
+        {3, 2.2},  // BNB Offbeam
+        {4, 10.1}, // NuMI Offbeam
+    };
+
+    /*double vetoOffset = 4;*/
+    double activeGate = gateLength[gateType] /*- vetoOffset*/;
+
+    double relFlashTime = flashTime + gateDiff / 1000. /*- vetoOffset*/;
+    mf::LogInfo("FilterCRTPMTMatching::flashInTime") << "Gate Diff " << gateDiff / 1000 << " Ftime+gateDiff " << flashTime + gateDiff / 1000. << " " << activeGate;
+
+    return ((relFlashTime > flashTimeCut) && (relFlashTime < activeGate));
+}
+
+icarus::crt::MatchedCRT CRTHitmatched(const double &flashTime, const double flashpos[3], std::vector<art::Ptr<sbn::crt::CRTHit>> &crtHits, const double &interval)
+{
+
+    std::vector<icarus::crt::CRTPMT> enteringCRTHits;
+    std::vector<icarus::crt::CRTPMT> exitingCRTHits;
+
+    int hasCRTHit = 0;
+    int topen = 0, topex = 0, sideen = 0, sideex = 0;
+    for (auto const crtHit : crtHits)
+    {
+        double tof = crtHit->ts1_ns / 1e3 - flashTime;
+        double distance = sqrt(pow(flashpos[0] - crtHit->x_pos, 2) + pow(flashpos[1] - crtHit->y_pos, 2) + pow(flashpos[2] - crtHit->z_pos, 2));
+        // std::cout<<"TOF "<<tof<<" "<<crtHit->ts1_ns<<" "<<crtHit->plane<<std::endl;
+        if (tof < 0 && abs(tof) < interval)
+        {
+            if (crtHit->plane > 36)
+                sideen++;
+            else
+                topen++;
+            CRTPMT m_match = {tof, distance, crtHit};
+            enteringCRTHits.push_back(m_match);
```

Comment:
I tried your suggestion, but the compiler is not a fan. I will try to fix it later on.

Reply:
Never mind then. It may be that `emplace_back()` does not cope well with POD(Plain Old Data) structures which don't have an explicit constructor. My bad.